core/model/src/main/java/org/eclipse/rdf4j/model/util/GraphComparisons.java

```diff
@@ -0,0 +1,415 @@
+
+	private static final HashFunction hashFunction = Hashing.murmur3_128();
+
+	private static final HashCode initialHashCode = hashFunction.hashString("", Charsets.UTF_8);
+	private static final HashCode outgoing = hashFunction.hashString("+", Charsets.UTF_8);
+	private static final HashCode incoming = hashFunction.hashString("-", Charsets.UTF_8);
+	private static final HashCode distinguisher = hashFunction.hashString("@", Charsets.UTF_8);
+
+	/**
+	 * Compares two RDF models, and returns <tt>true</tt> if they consist of isomorphic graphs and the isomorphic graph
+	 * identifiers map 1:1 to each other. RDF graphs are isomorphic graphs if statements from one graphs can be mapped
+	 * 1:1 on to statements in the other graphs. In this mapping, blank nodes are not considered mapped when having an
+	 * identical internal id, but are mapped from one graph to the other by looking at the statements in which the blank
+	 * nodes occur. A Model can consist of more than one graph (denoted by context identifiers). Two models are
+	 * considered isomorphic if for each of the graphs in one model, an isomorphic graph exists in the other model, and
+	 * the context identifiers of these graphs are identical.
+	 * 
+	 * @implNote The algorithm used by this comparison is a depth-first search for an iso-canonical blank node mapping
+	 *           for each model, and using that as a basis for comparison. The algorithm is described in detail in:
+	 *           Hogan, A. (2017). Canonical forms for isomorphic and equivalent RDF graphs: algorithms for leaning and
+	 *           labelling blank nodes. ACM Transactions on the Web (TWEB), 11(4), 1-62.
+	 *
+	 * @see <a href="http://www.w3.org/TR/rdf11-concepts/#graph-isomorphism">RDF Concepts &amp; Abstract Syntax, section
+	 *      3.6 (Graph Comparison)</a>
+	 * @see http://aidanhogan.com/docs/rdf-canonicalisation.pdf
+	 * 
+	 */
+	public static boolean isomorphic(Model model1, Model model2) {
+		if (model1 == model2) {
+			return true;
+		}
+
+		if (model1.size() != model2.size()) {
+			return false;
+		}
+
+		if (model1.contexts().size() != model2.contexts().size()) {
+			return false;
+		}
+
+		if (model1.contexts().size() > 1) {
+			// model contains more than one context (including the null context). We compare per individual context.
+			for (Resource context : model1.contexts()) {
+				if (context != null && context.isBNode()) {
+					// We currently do not handle mapping of blank nodes used as context identifiers.
+					logger.warn(
+							"isomorphism detection can not map blank nodes used as context identifiers. Comparison may give inaccurate results",
+							context
+					);
+				}
+
+				Model contextInlModel1 = model1.filter(null, null, null, context);
+				Model contextInModel2 = model2.filter(null, null, null, context);
+				if (contextInlModel1.size() != contextInModel2.size()) {
+					return false;
+				}
+				final Model canonicalizedContext1 = isoCanonicalize(contextInlModel1);
```

Comment:
FYI. Final variables inside methods don't make any difference for performance.

Reply:
I tend to use final as a statement of intent rather than (purely) for performance reasons - plus in code where there's a lot juggling of (in this case) Model objects going on, it is helpful to make sure a reference is not overridden later on down.